### PR TITLE
feature: nanocl state file

### PR DESCRIPTION
**Pull request by leon3s**
Hey, wanted to help you out to write the `yaml` for nanocl.
I also used `working_dir` value inside docker compose to avoid the use of `cd` !

Note: I need to make a fix for the env variable so we can set default value for unexisting env. Right now nanocl wont let you apply a state if you use and empty variable..

### CODE REVIEW

**Comment by darioalessandro** [comment]
@leon3s this is awesome! I am installing `nanocl` and will test the generated state file.

File: bot/src/main.rs [hunk 1]
```diff
@@ -26,7 +26,7 @@ async fn main() {
     }
 
     for client in clients {
-        client.await;
+        let _ = client.await;
```
Comment (darioalessandro) [inline bot/src/main.rs:29]: yes! I know this warning was super annoying

**Comment by leon3s** [comment]
You should wait for my fix with the env varible i think. i didn't filled everything correctly i think.
Also i should add resource so you can have real domain name even in development. I wasn't able to test the project with the docker compose. A 404 html page appear but the request return a 200.

**Comment by darioalessandro** [comment]
Thank you @leon3s I really appreciate it!

**Comment by leon3s** [comment]
So i fully finished the `yaml` config. And this is the page i'm seing even with docker compose:
image

I did encounter two problems, i use [liquid](https://docs.rs/liquid/latest/liquid/) to replace variable by value with this syntax:
```
${{ Envs.MY_ENV }}
```
But the library doesn't allow non existing key to have default value the key have to be set as null.
The workarround i choose is to add `dotenv` on the cli and so you can specify blank variable in a `.env`

The second problem is right now i don't really have equivalent of `depend_on`. I'll add a feature for that, maybe something like `init_container` as i could read [there](https://stackoverflow.com/questions/49368047/what-is-the-equivalent-for-depends-on-in-kubernetes)

File: docker/StateFile.yml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        - trunk serve --address 0.0.0.0 --port 8081
+      HostConfig:
+        Dns:
+          - ${{ Namespaces.global.Gateway }}
+          - 1.1.1.1
+        Binds:
+          - ../:/app
+          - ${{ Envs.HOME }}/.cargo/registry:/usr/local/cargo/registry
+      Env:
+      - ACTIX_HOST=localhost
```
Comment (griffobeid) [inline docker/StateFile.yml:46]: Would it be possible for this to use the DNS `api.zoom.rs` instead?

File: docker/StateFile.yml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      - ACTIX_PORT=8080
+      - ACTIX_UI_BACKEND_URL=ws://localhost:8080
```
Comment (griffobeid) [inline docker/StateFile.yml:48]: Same could this use DNS instead?

File: docker/StateFile.yml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      - TRUNK_SERVE_PORT=8081
+      - ENABLE_OAUTH=false
+      - LOGIN_URL=http://localhost:8080/login
```
Comment (griffobeid) [inline docker/StateFile.yml:51]: Same could this use DNS instead?

**Review by griffobeid — COMMENTED** [review]
@leon3s Thanks for the work on this, this is awesome! I'm going to set up an Ubuntu VM on my Macbook so I can test this out. Just one comment about using DNS instead of localhost if possible so that it could work on a cloud environment

**Comment by leon3s** [comment]
Oups my bad! Yeah ofc, one great thing about nanocl is to be able to have a dev environnement really close to the production!

If you want to avoid to install a Virtual Machine you can use the docker image of the CLI:
```Dockerfile
FROM ghcr.io/nxthat/nanocl:0.8.1 as builder

FROM debian:latest

COPY --from=builder /bin/nanocl /bin/nanocl

RUN groupadd nanocl

ENTRYPOINT ["/bin/nanocl"]
```

So you should be able to run it this way:
```sh
docker run -v /run/nanocl:/run/nanocl -v /run/docker.sock:/run/docker.sock nanocl:latest version
```

And you can create an alias on top of it:

```sh
alias nanocl="docker run -v /run/nanocl:/run/nanocl -v /run/docker.sock:/run/docker.sock nanocl:latest"
```

It's up to you!

**Comment by griffobeid** [comment]
Nice that's better than a VM 🙏

**Comment by leon3s** [comment]
You need to run a `nanocl install` then you should be able to work with it!

**Comment by griffobeid** [comment]
Okay, I tried the install and it is giving me a network error. I'm in Budapest on shitty WiFi so it may be on my end here is the full response:

```sh
Installing Nanocl components on your system
Using default gateway: 172.17.0.2
Using default hostname: 1002c5ebfdea
Creating system network
Nanocl system network: Error trying to connect: Connection refused (os error 111)
```

Also tried with `--net=host` on the docker run with the same results. A MacOS binary would be the best I think, I'll try to compile it locally

**Comment by griffobeid** [comment]
Giving `cargo install nanocl` a shot 🤞

**Comment by leon3s** [comment]
This is because you are using docker desktop ? do a `docker context ls` the docker.sock is not under `/run/docker.sock` in such case. I can help you on discord if you want

**Comment by griffobeid** [comment]
So `cargo install nanocl` did install the nanocl cli, but `nanocl install` is giving a `No such file or directory` error, so it must be trying to access some file that exists on Ubuntu but not MacOS I'm guessing. 

Here is the output of `nanocl version`

```sh
=== [nanocli] ===
Arch: x86_64
Channel: stable
Version: 0.7.0
Commit ID:
/run/nanocl/nanocl.sock: Connector has been disconnected
```

**Comment by leon3s** [comment]
Yeah as i said it's because your docker.sock is not under `/run/docker.sock`

**Comment by leon3s** [comment]
The lasted version wasn't published on crates.io i just published it. You can retry `cargo install nanocl`
and then to install `nanocl install --docker-host ~/.docker/desktop/docker.sock`

**Comment by leon3s** [comment]
Also you can try [multipass](https://multipass.run/) to quickly speedup ubuntu vm

**Comment by leon3s** [comment]
I did send you an email with my discord if you want, actually it wont work because i didn't mount `~/.docker/desktop/docker.sock` you need to tweak the installer to make it happen.. I think the easier move is to create a symbolic link from from `~/.docker/desktop/docker.sock` to `/run/docker.sock`.

You can see what does the installer [there](https://github.com/nxthat/nanocl/blob/nightly/installer.yml)
And you can download it tweak it to add the `/home/{user}/.docker/desktop/docker.sock` in the binds
```yml
        Binds:
          - /home/{user}/.docker/desktop/docker.sock:/run/nanocl
          - /run/docker.sock:/run/docker.sock
          - ${{ state_dir }}:${{ state_dir }}
          - ${{ conf_dir }}:${{ conf_dir }}
```

**Comment by leon3s** [comment]
It's kinda tricky on MacOS and that because docker didn't release a binary for the new ships so you have different installation context

**Comment by griffobeid** [comment]
No worries! I just want to help find the least path of resistance here with you for MacOS, thanks for working with me on it 😄 my Macbook is an older intel so that might help, also I noticed the sock path is at `~/.docker/run/docker.sock`

**Comment by leon3s** [comment]
@griffobeid Thank for taking the time to solve this with me!
So the quickest way will be to create a symbolic link from `~/.docker/run/docker.sock` to `/run/docker.sock` with `ln -s` if i remember well. But that wont persist after reboots.
The second solution is to use your own installation template copy the following content into a `installer.yml`:

```yml
Kind: Deployment
ApiVersion: v0.8

Namespace: system

Cargoes:
  - Name: nmetrics
    Container:
      Image: ghcr.io/nxthat/metrsd:0.2.0
      Tty: true
      Env:
        - TZ=Europe/Paris
      Cmd:
        - --hosts
        - unix:///run/nanocl/metrics.sock
      HostConfig:
        NetworkMode: system
        Binds:
          - /run/nanocl:/run/nanocl

  - Name: nproxy
    Container:
      Image: ghcr.io/nxthat/nproxy:1.24.0.2
      Tty: true
      Env:
        - TZ=Europe/Paris
      HostConfig:
        NetworkMode: host
        Binds:
          - /run:/run
          - ${{ state_dir }}/proxy/logs:/var/log/nginx/access
          - ${{ state_dir }}/proxy/certs:/etc/nginx/certs
          - ${{ state_dir }}/proxy/letsencrypt:/etc/letsencrypt
          - ${{ state_dir }}/proxy/conf.d:/etc/nginx/conf.d
          - ${{ state_dir }}/proxy/html:/usr/share/nginx/html
          - ${{ state_dir }}/proxy/sites-enabled:/etc/nginx/sites-enabled
          - ${{ state_dir }}/proxy/streams-enabled:/etc/nginx/streams-enabled

  - Name: ncdproxy
    Container:
      Image: ghcr.io/nxthat/ncdproxy:0.5.0
      Tty: true
      Env:
        - TZ=Europe/Paris
      HostConfig:
        NetworkMode: system
        Binds:
          - /run/nanocl:/run/nanocl
          - ${{ state_dir }}/proxy/conf.d:/etc/nginx/conf.d
          - ${{ state_dir }}/proxy/logs:/var/log/nginx/access
          - ${{ state_dir }}/proxy/sites-enabled:/etc/nginx/sites-enabled
          - ${{ state_dir }}/proxy/streams-enabled:/etc/nginx/streams-enabled

  - Name: ndns
    Container:
      Env:
        - TZ=Europe/Paris
      Image: ghcr.io/nxthat/ndns:2.87.0
      HostConfig:
        NetworkMode: host
        Binds:
          - ${{ state_dir }}/dns:/opt/dns

  - Name: ncddns
    Container:
      Env:
        - TZ=Europe/Paris
      Image: ghcr.io/nxthat/ncddns:0.2.0
      Tty: true
      Cmd:
        - --conf-dir
        - /opt/dns
        - --dns
        - 1.1.1.1
        - --dns
        - 1.0.0.1
      HostConfig:
        NetworkMode: system
        Binds:
          - /run/nanocl:/run/nanocl
          - ${{ state_dir }}/dns:/opt/dns

  - Name: nstore
    Container:
      Image: cockroachdb/cockroach:v22.2.7
      Tty: true
      Hostname: nstore.nanocl.internal
      Env:
        - TZ=Europe/Paris
      Cmd:
        - start-single-node
        - --insecure
        - --listen-addr=:26257
        - --advertise-addr=${{ advertise_addr }}:26257
      HostConfig:
        NetworkMode: system
        Binds:
          - ${{ state_dir }}/store/ca:/ca
          - ${{ state_dir }}/store/certs:/certs
          - ${{ state_dir }}/store/data:/cockroach/cockroach-data

  - Name: ndaemon
    Container:
      Image: ghcr.io/nxthat/nanocld:0.8.0
      Tty: true
      Hostname: ndaemon.nanocl.internal
      Cmd:
        - --state-dir
        - ${{ state_dir }}
        - --conf-dir
        - ${{ conf_dir }}
        - --docker-host
        - ${{ docker_host }}
        - --gateway
        - ${{ gateway }}
        - --hostname
        - ${{ hostname }}
        - --advertise-addr
        - ${{ advertise_addr }}
        - --gid
        - "${{ gid }}"
      Env:
        - TZ=Europe/Paris
        - NANOCL_GID=${{ gid }}
      HostConfig:
        NetworkMode: system
        Binds:
          - /run/nanocl:/run/nanocl
          - /home/{user}/.docker/run/docker.sock:/home/{user}/.docker/run/docker.sock # Replace {user} with your username
          - ${{ state_dir }}:${{ state_dir }}
          - ${{ conf_dir }}:${{ conf_dir }}
```

You need to replace `{user}` with your username
And then running the installation template this way:
```sh
nanocl install -t installer.yml --docker-host /home/{user}/.docker/run/docker.sock # Replace {user} with your username
```

You cannot use the `~` you need to specify the full path of the `docker.sock`

This should work, if it doesn't i'll setup a MacOs Virtual Machine and see what is going wrong!

**Comment by griffobeid** [comment]
Hey @ leon3s I'm not too concerned about the MacOS support atm, we can merge this if it works on Ubuntu. I'm testing on my Ubuntu 22.04 machine. I went through all the setup and tutorials on my machine and that is all working well! 

When I go test the StateFile.yml in this pull request nanocl is returning 

```
/run/nanocl/nanocl.sock: Connector has been disconnected
```

Any ideas?

**Comment by leon3s** [comment]
@griffobeid that's mean the daemon is down or maybe it's a permission problem on the unix socket.
What is the output of `ls -l /run/nanocl` or `docker logs ndaemon.system.c` ?

**Comment by griffobeid** [comment]
`docker logs ndaemon.system.c` is outputing this:

```
[2023-06-13T20:38:33Z INFO ] Connecting to store at: nstore.nanocl.internal:26257
[2023-06-13T20:38:37Z INFO ] Store connected
[2023-06-13T20:38:37Z INFO ] Running migrations
[2023-06-13T20:38:37Z INFO ] Migrations done
[2023-06-13T20:38:37Z INFO ] Syncing existing container
[2023-06-13T20:38:37Z DEBUG] updating cargo ncddns in namespace system
[2023-06-13T20:38:38Z DEBUG] updating cargo ncdproxy in namespace system
[2023-06-13T20:38:38Z DEBUG] updating cargo nmetrics in namespace system
[2023-06-13T20:38:38Z INFO ] Container synced
[2023-06-13T20:38:38Z DEBUG] set_unix_permission
[2023-06-13T20:38:38Z DEBUG] watching change of: /run/nanocl
[2023-06-13T20:38:38Z INFO ] Preparing server
[2023-06-13T20:38:38Z INFO ] Listening on unix:///run/nanocl/nanocl.sock
[2023-06-13T20:38:38Z DEBUG] event: Event { kind: Remove(File), paths: ["/run/nanocl/nanocl.sock"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
[2023-06-13T20:38:38Z INFO ] Server ready
[2023-06-13T20:38:38Z DEBUG] event: Event { kind: Create(File), paths: ["/run/nanocl/nanocl.sock"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
[2023-06-13T20:38:38Z DEBUG] change detected, change permission of /run/nanocl/nanocl.sock
[2023-06-13T20:38:38Z DEBUG] set_unix_permission done
```

`ls -l /run/nanocl` is giving permission denied, with `sudo ls -l /run/nanocl` 

```
total 0
srwxrwx--- 1 root nanocl 0 Jun 13 16:38 dns.sock
srwxrwx--- 1 root nanocl 0 Jun 13 16:38 metrics.sock
srwxrwx--- 1 root nanocl 0 Jun 13 16:38 nanocl.sock
srwxrwx--- 1 root nanocl 0 Jun 13 16:38 proxy.sock
```

**Comment by leon3s** [comment]
Your current shell is not updated with the nanocl group then, try a `newgrp nanocl`

**Comment by leon3s** [comment]
@griffobeid does it worked ? seems like everything is running fine from the logs

**Comment by griffobeid** [comment]
Yes 💯

**Comment by griffobeid** [comment]
Now I'm having trouble with the dns, the server and dbmate are unable to connect to the database, looks like the ip is resolving to `172.24.0.1` instead of the pod IP

**Comment by leon3s** [comment]
With the config from this branch ?

**Comment by griffobeid** [comment]
Yes using the config here. I checked the logs of the DNS docker and there are some warnings 

```
[2023-06-13T20:38:27Z INFO ] ncddns v0.2.0
[2023-06-13T20:38:27Z INFO ] Ensuring a minimal dnsmasq config inside /opt/dns
[2023-06-13T20:38:27Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:27Z INFO ] Minimal dnsmasq config is ensured
[2023-06-13T20:38:27Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:27Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:29Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:29Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:29Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:31Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:31Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:31Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:33Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:33Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:33Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:35Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:35Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:35Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:37Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:37Z WARN ] Unable to Subscribe to nanocl daemon events: /run/nanocl/nanocl.sock: Connector has been disconnected
[2023-06-13T20:38:37Z WARN ] Unsubscribed from nanocl daemon events, retrying to subscribe in 2 seconds
[2023-06-13T20:38:39Z INFO ] Subscribing to nanocl daemon events..
[2023-06-13T20:38:39Z INFO ] Subscribed to nanocl daemon events
[2023-06-13T20:38:39Z INFO ] DnsRule already exists. Skipping.
```

File: docker/StateFile.yml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          - ${{ Envs.HOME }}/.cargo/registry:/usr/local/cargo/registry
+      Env:
+      - ACTIX_HOST=api.zoom.rs
+      - ACTIX_PORT=8080
+      - ACTIX_UI_BACKEND_URL=ws://api.zoom.rs
+      - TRUNK_SERVE_PORT=8081
+      - ENABLE_OAUTH=false
+      - LOGIN_URL=http://api.zoom.rs/login
+      - RUSTFLAGS=--cfg=web_sys_unstable_apis
+      - RUST_BACKTRACE=1
```
Comment (griffobeid) [inline docker/StateFile.yml:53]: Does this need to be indented once?

Comment (leon3s) [reply docker/StateFile.yml:53]: Both are valid, you can do a `nanocl cargo inspect {cargo_name}` to verify if it's correctly set

**Comment by leon3s** [comment]
It's just a warning that say the daemon is not reachable yet, but it's time for the daemon to boot so it's normal, the end of the logs show that the dns successfully connected. The database is binded with a ProxyRule to the gateway of the network so the dns resolution is correct. But you may have port already open on the same port that the database want to use.

**Pull request closed without merging**